psi4/src/psi4/libsapt_solver/exch-ind30.cc

```diff
@@ -314,5 +319,305 @@ double SAPT2p3::exch_ind30_3(double **sBS) {
+
+    // C1's and C2's from D's and C's.
+    // C1's are C times D diagonal blocks.
+    // C2's are times off-diagonal blocks.
+    auto C1a = std::make_shared<Matrix>("C1a", nn, na);
+    auto C1b = std::make_shared<Matrix>("C1b", nn, nb);
+    auto C2a = std::make_shared<Matrix>("C2a", nn, na);
+    auto C2b = std::make_shared<Matrix>("C2b", nn, nb);
+
+    C_DGEMM('N', 'N', nn, na, na, 1.0, CoccA_->pointer()[0], na, &Dp[0][0], na + nb, 0.0,
+            C1a->pointer()[0], na);
+    C_DGEMM('N', 'N', nn, nb, nb, 1.0, CoccB_->pointer()[0], nb, &Dp[na][na], na + nb, 0.0,
+            C1b->pointer()[0], nb);
+    C_DGEMM('N', 'N', nn, na, nb, 1.0, CoccB_->pointer()[0], nb, &Dp[na][0], na + nb, 0.0,
+            C2a->pointer()[0], na);
+    C_DGEMM('N', 'N', nn, nb, na, 1.0, CoccA_->pointer()[0], na, &Dp[0][na], na + nb, 0.0,
+            C2b->pointer()[0], nb);
+
+    // Coeffs for all occupied
+    std::vector<std::shared_ptr<Matrix> > hold_these;
+    hold_these.push_back(CoccA_);
+    hold_these.push_back(CoccB_);
+
+    auto Cocc0AB = linalg::horzcat(hold_these);
+    hold_these.clear();
+
+    // Half transform D_ia and D_ib for JK
+    auto D_Ni_a = std::make_shared<Matrix>("D_Ni_a", nn, na + nb);
+    auto D_Ni_b = std::make_shared<Matrix>("D_Ni_b", nn, na + nb);
+
+    C_DGEMM('N', 'N', nn, na + nb, na, 1.0, CoccA_->pointer()[0], na, &Dp[0][0], na + nb, 0.0,
+            D_Ni_a->pointer()[0], na + nb);
+    C_DGEMM('N', 'N', nn, na + nb, nb, 1.0, CoccB_->pointer()[0], nb, &Dp[na][0], na + nb, 0.0,
+            D_Ni_b->pointer()[0], na + nb);
+
+    // Global JK object
+    std::shared_ptr<JK> jk_;
+
+    // Make JK's
+    jk_ = JK::build_JK(basisset_, get_basisset("DF_BASIS_SCF"), options_, false, mem_);
+    jk_->set_memory(mem_);
+    jk_->set_do_J(true);
+    jk_->set_do_K(true);
+    jk_->initialize();
+    jk_->print_header();
+
+    std::vector<SharedMatrix>& Cl = jk_->C_left();
+    std::vector<SharedMatrix>& Cr = jk_->C_right();
+    const std::vector<SharedMatrix>& J = jk_->J();
+    const std::vector<SharedMatrix>& K = jk_->K();
+
+    Cl.clear();
+    Cr.clear();
+    Cl.push_back(Cocc0AB);
+    Cr.push_back(D_Ni_a);
+    Cl.push_back(Cocc0AB);
+    Cr.push_back(D_Ni_b);
+    jk_->compute();
+
+    std::shared_ptr<Matrix> J_D_ia = J[0];
+    std::shared_ptr<Matrix> K_D_ia = K[0];
+
+    std::shared_ptr<Matrix> J_D_ib = J[1];
+    std::shared_ptr<Matrix> K_D_ib = K[1];
+
+    // Finish D_ia and D_ib transformation to make tilded C's
+    auto D_ia = linalg::doublet(Cocc0AB, D_Ni_a, false, true);
+    auto D_ib = linalg::doublet(Cocc0AB, D_Ni_b, false, true);
+
+    auto Ct_Kr = linalg::triplet(D_ib, Smat_, CvirA_, false, false, false);
+    Ct_Kr->scale(-1);
+    Ct_Kr->add(CvirA_);
+    auto Ct_Ks = linalg::triplet(D_ia, Smat_, CvirB_, false, false, false);
+    Ct_Ks->scale(-1);
+    Ct_Ks->add(CvirB_);
+
+    // Make omega cores
+    std::shared_ptr<Matrix> AJK(J_D_ia->clone());
+    AJK->zero();
+    AJK->add(J_D_ia);
```

Comment:
Recurring issue: I don't understand why you clone a matrix, zero it, and then add the matrix you just cloned. Just clone it.

Reply:
My bad. Fixed throughout.